File: reference/goathal/conformance.py
"""
D.1 conformance suite runner (PROTOCOL layer — device-agnostic).

Runs Spec D criteria D1-D8 against ANY GoatBackend via the trait only. It never reads
class_id to decide behavior; a backend's class is opaque. Returns a structured report.
This is the objective gate for Backend Bounty award at Stage 1 and (via the
maintenance-stream CI predicate, D.2) for continuation.

`backend_factory(node_seed) -> GoatBackend` produces an independent "node".
"""
from dataclasses import dataclass
from typing import Dict
from .types import Task, TaskResult, ExecPolicy, Preempt, DetKind
from .commit import commit


@dataclass
class CriterionResult:
    passed: bool
    detail: str


@dataclass
class ConformanceReport:
    results: Dict[str, CriterionResult]

    @property
    def all_passed(self) -> bool:
        return all(r.passed for r in self.results.values())


def _l_inf(a, b):
    return max(abs(x - y) for x, y in zip(a, b)) if a and b else 0


def run_conformance(backend_factory, node_count: int = 25) -> ConformanceReport:
    r: Dict[str, CriterionResult] = {}
    build = "build-1"
    task = Task(task_class_id=10, engine_build_id=build, payload=b"opaque-corpus-0",
                seed=1, determinism_bound=10.0)
    ref = backend_factory(0)
    dev = ref.enumerate_devices()[0]

    # D1 interface completeness
    required = ("enumerate_devices", "benchmark", "determinism_profile", "load",
                "execute", "commit", "preempt", "telemetry", "enforce_envelope",
                "idle_signals")
    missing = [m for m in required if not callable(getattr(ref, m, None))]
    r["D1_interface"] = CriterionResult(not missing, f"missing={missing}")

    # D2 capability honesty: benchmark reproduces within tolerance across N nodes
    rates = [backend_factory(i).benchmark(dev).task_class_caps[0].measured_gcu_rate
             for i in range(node_count)]
    median = sorted(rates)[len(rates) // 2]
    spread = max(abs(x - median) / median for x in rates)
    r["D2_capability_honesty"] = CriterionResult(spread <= 0.02, f"rel_spread={spread:.4f}")

    # D3 determinism declaration holds empirically
    prof = ref.determinism_profile(dev, 10)
    outs = [backend_factory(0).execute(backend_factory(0).load(dev, build), task,
                                       ExecPolicy(power_cap_w=10_000)).vector
            for _ in range(5)]
    worst = max(_l_inf(outs[0], o) for o in outs[1:]) if len(outs) > 1 else 0
    d3_ok = worst <= (0.0 if prof.kind == DetKind.EXACT else prof.bound)
    r["D3_determinism"] = CriterionResult(
        d3_ok, f"kind={prof.kind.name} worst_l_inf={worst} bound={prof.bound}")

    # D4 canonical commitment: byte-identical commit across independent nodes
    res_a = backend_factory(3).execute(backend_factory(3).load(dev, build), task,
                                       ExecPolicy(power_cap_w=10_000))
    res_b = backend_factory(9).execute(backend_factory(9).load(dev, build), task,
                                       ExecPolicy(power_cap_w=10_000))
    d4_ok = commit(res_a).digest == commit(res_b).digest
    r["D4_canonical_commit"] = CriterionResult(d4_ok, f"match={d4_ok}")

    # D5 preemption: a preempt request yields SavedState (not a full result), p95 declared
    p95 = getattr(ref, "preempt_p95_ms", None)
    saved = ref.execute(ref.load(dev, build), task, ExecPolicy(power_cap_w=10_000),
                        preempt=Preempt(requested=True))
    d5_ok = (p95 is not None) and (not isinstance(saved, TaskResult)) and hasattr(saved, "progress_chunks")
    r["D5_preemption"] = CriterionResult(bool(d5_ok), f"p95_ms={p95} yielded_savedstate={not isinstance(saved, TaskResult)}")

    # D6 envelope enforcement: peak power under load never exceeds the cap
    cap_w = 50
    b6 = backend_factory(0)
    b6.enforce_envelope(dev, cap_w)
    b6.execute(b6.load(dev, build), task, ExecPolicy(power_cap_w=cap_w))
    peak = b6.peak_power_w
    r["D6_envelope"] = CriterionResult(0 < peak <= cap_w, f"peak_power_w={peak} cap={cap_w}")

    # D7 telemetry fidelity: values in range, util normalized
    t = backend_factory(0).telemetry(dev)
    d7_ok = t.power_w >= 0 and 0.0 <= t.util <= 1.0 and t.mem_used_mb >= 0
    r["D7_telemetry"] = CriterionResult(d7_ok, f"power={t.power_w} util={t.util}")

    # D8 neutrality (behavioral half): commit is a function of task semantics only,
    # never device identity -> an identical result twin commits identically.
    twin = TaskResult(task_class_id=res_a.task_class_id, tokens=res_a.tokens,
                      vector=res_a.vector, engine_build_id=res_a.engine_build_id)
    r["D8_neutrality_behavioral"] = CriterionResult(
        commit(res_a).digest == commit(twin).digest,
        "commit depends on task semantics only, not device identity")

    return ConformanceReport(results=r)

File: goatcoin-rs/crates/goat-protocol/src/conformance.rs
//! D.1 conformance suite runner (device-agnostic). Runs D1-D8 against ANY GoatBackend via
//! the trait only; never reads class_id to decide behavior. D-1: D6 uses a peak-power
//! observable and enforce_envelope binds over the per-task policy.

use std::collections::BTreeMap;

use crate::backend::GoatBackend;
use crate::commit::commit;
use crate::types::{DetKind, ExecOutcome, ExecPolicy, Preempt, Task, TaskResult};

pub struct ConformanceReport {
    pub results: BTreeMap<String, (bool, String)>,
}

impl ConformanceReport {
    pub fn all_passed(&self) -> bool {
        self.results.values().all(|(ok, _)| *ok)
    }
}

fn l_inf(a: &[i64], b: &[i64]) -> i64 {
    if a.len() != b.len() || a.is_empty() {
        return if a.is_empty() && b.is_empty() {
            0
        } else {
            i64::MAX
        };
    }
    a.iter()
        .zip(b)
        .map(|(x, y)| (x - y).abs())
        .max()
        .unwrap_or(0)
}

/// `factory(node_seed) -> Backend`. D2 uses independent nodes; D4 uses fresh nodes.
pub fn run_conformance<B: GoatBackend>(
    factory: &dyn Fn(u64) -> B,
    node_count: u64,
) -> ConformanceReport {
    let mut r = BTreeMap::new();
    let build = "build-1";
    let task = Task {
        task_class_id: 10,
        engine_build_id: build.into(),
        payload: b"opaque-corpus-0".to_vec(),
        seed: 1,
        determinism_bound: 10.0,
    };
    let refb = factory(0);
    let dev = refb.enumerate_devices()[0].clone();

    // D1 interface completeness is guaranteed by the trait; record trivially true.
    r.insert("D1_interface".into(), (true, "trait-guaranteed".into()));

    // D2 capability honesty: benchmark reproduces within tolerance across N nodes
    let rates: Vec<f64> = (0..node_count)
        .map(|i| factory(i).benchmark(&dev).task_class_caps[0].measured_gcu_rate)
        .collect();
    let mut sorted = rates.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let median = sorted[sorted.len() / 2];
    let spread = rates
        .iter()
        .map(|x| (x - median).abs() / median)
        .fold(0.0, f64::max);
    r.insert(
        "D2_capability_honesty".into(),
        (spread <= 0.02, format!("rel_spread={spread:.4}")),
    );

    // D3 determinism declaration holds empirically
    let prof = refb.determinism_profile(&dev, 10);
    let mut outs: Vec<Vec<i64>> = Vec::new();
    for _ in 0..5 {
        let mut b = factory(0);
        if let ExecOutcome::Completed(res) = b.execute(
            &dev,
            &task,
            ExecPolicy {
                power_cap_w: 10_000,
            },
            Preempt::default(),
        ) {
            outs.push(res.vector);
        }
    }
    let worst = outs
        .iter()
        .skip(1)
        .map(|o| l_inf(&outs[0], o))
        .max()
        .unwrap_or(0);
    let d3_ok = if prof.kind == DetKind::Exact {
        worst == 0
    } else {
        worst as f64 <= prof.bound
    };
    r.insert(
        "D3_determinism".into(),
        (
            d3_ok,
            format!("kind={:?} worst={worst} bound={}", prof.kind, prof.bound),
        ),
    );

    // D4 canonical commitment: byte-identical commit across independent nodes
    let run = |seed: u64| -> Option<TaskResult> {
        let mut b = factory(seed);
        match b.execute(
            &dev,
            &task,
            ExecPolicy {
                power_cap_w: 10_000,
            },
            Preempt::default(),
        ) {
            ExecOutcome::Completed(res) => Some(res),
            _ => None,
        }
    };
    let d4_ok = match (run(3), run(9)) {
        (Some(x), Some(y)) => commit(&x) == commit(&y),
        _ => false,
    };
    r.insert(
        "D4_canonical_commit".into(),
        (d4_ok, format!("match={d4_ok}")),
    );

    // D5 preemption: a preempt request yields a SavedState, not a full result
    let mut b5 = factory(0);
    let saved = b5.execute(
        &dev,
        &task,
        ExecPolicy {
            power_cap_w: 10_000,
        },
        Preempt { requested: true },
    );
    let d5_ok = matches!(saved, ExecOutcome::Preempted(_)) && refb.preempt_p95_ms() > 0;
    r.insert(
        "D5_preemption".into(),
        (
            d5_ok,
            format!(
                "p95={} yielded={}",
                refb.preempt_p95_ms(),
                matches!(saved, ExecOutcome::Preempted(_))
            ),
        ),
    );

    // D6 envelope enforcement: peak power under load never exceeds the cap (D-1)
    let cap = 50u32;
    let mut b6 = factory(0);
    b6.enforce_envelope(&dev, cap);
    let _ = b6.execute(
        &dev,
        &task,
        ExecPolicy { power_cap_w: cap },
        Preempt::default(),
    );
    let peak = b6.peak_power_w();
    r.insert(
        "D6_envelope".into(),
        (
            peak > 0.0 && peak <= cap as f64,
            format!("peak={peak} cap={cap}"),
        ),
    );

    // D7 telemetry fidelity
    let t = factory(0).telemetry(&dev);
    r.insert(
        "D7_telemetry".into(),
        (
            t.power_w >= 0.0 && (0.0..=1.0).contains(&t.util),
            format!("power={} util={}", t.power_w, t.util),
        ),
    );

    // D8 neutrality (behavioral half): commit is a function of task semantics only
    let res = run(3).unwrap();
    let twin = TaskResult {
        task_class_id: res.task_class_id,
        tokens: res.tokens.clone(),
        vector: res.vector.clone(),
        engine_build_id: res.engine_build_id.clone(),
    };
    r.insert(
        "D8_neutrality_behavioral".into(),
        (
            commit(&res) == commit(&twin),
            "commit depends on task semantics only".into(),
        ),
    );

    ConformanceReport { results: r }
}
